src/sentry/search/snuba/backend.py

```diff
@@ -219,43 +231,73 @@ def _query(self, project, retention_window_start, group_queryset, tags, environm
-        # queryset might return a *huge* number of groups. In this case, we
-        # probably *don't* want to pass candidates down to Snuba, and rather we
-        # want Snuba to do all the filtering/sorting it can and *then* apply
-        # this queryset to the results from Snuba.
-        #
-        # However, if this did filter down the number of groups significantly,
-        # then passing in candidates is, of course, valuable.
-        #
-        # Should we decide which way to handle it based on the number of
-        # group_ids, the number of hashes? Or should we just always start the
-        # query with Snuba? Something else?
-        candidate_group_ids = list(group_queryset.values_list('id', flat=True))
-
-        sort, extra_aggregations, calculate_cursor_for_group = sort_strategies[sort_by]
-
-        group_data = do_search(
+        candidate_hashes = None
+        if MAX_PRE_SNUBA_CANDIDATES:
+            # If the query didn't include anything to significantly filter
+            # down the number of group hashes at this point ('first_release', 'query',
+            # 'status', 'bookmarked_by', 'assigned_to', 'unassigned',
+            # 'subscribed_by', 'active_at_from', or 'active_at_to') then this
+            # queryset might return a *huge* number of groups. In this case, we
+            # probably *don't* want to pass candidates down to Snuba, and rather we
+            # want Snuba to do all the filtering/sorting it can and *then* apply
+            # this queryset to the results from Snuba.
+            #
+            # However, if this did filter down the number of groups significantly,
+            # then passing in candidates is, of course, valuable.
+
+            # {hash: group_id, ...}
+            candidate_hashes = dict(
+                GroupHash.objects.filter(
+                    group__in=group_queryset
+                ).values_list(
+                    'hash', 'group_id'
+                )[:MAX_PRE_SNUBA_CANDIDATES + 1]
+            )
+
+            if not candidate_hashes:
+                # no matches could possibly be found from this point on
+                return Paginator(Group.objects.none()).get_result()
+            elif len(candidate_hashes) > MAX_PRE_SNUBA_CANDIDATES:
+                # too many candidate hashes found, so we won't pass any down to Snuba
+                # and we'll do post-filtering instead
+                candidate_hashes = None
+
+        sort, extra_aggregations, score_fn = sort_strategies[sort_by]
+
+        # {group_id: group_score, ...}
+        snuba_groups = snuba_search(
             project_id=project.id,
             environment_id=environment and environment.id,
             tags=tags,
             start=start,
             end=end,
             sort=sort,
             extra_aggregations=extra_aggregations,
-            candidates=candidate_group_ids,
+            score_fn=score_fn,
+            candidate_hashes=candidate_hashes,
```

Comment:
Not that it really matters, but this length can be `MAX_PRE_SNUBA_CANDIDATES` + 1 which makes the name a little misleading.

Reply:
Hm, I don't follow. Is this comment on the right line? I'm not sure what the length has to do with anything at this point, also if it's ever `max + 1` we convert to `None`: https://github.com/getsentry/sentry/blob/261fe7deb4ffaf8bbab5ec8eb9d435e263bc78f8/src/sentry/search/snuba/backend.py#L260

Reply:
Oops, read this too fast. You're right — ignore this.